File: app/src/main/java/com/example/desafioandroid/feature/repository/RepositoryActivity.java
package com.example.desafioandroid.feature.repository;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.os.PersistableBundle;
import android.util.Log;
import android.widget.AbsListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.desafioandroid.R;
import com.example.desafioandroid.feature.endpoints.GetDataService;
import com.example.desafioandroid.feature.pulls.PullActivity;
import com.example.desafioandroid.feature.utilities.EndlessScroll;
import com.example.desafioandroid.feature.utilities.RetrofitClientInstance;
import com.example.desafioandroid.model.Base;
import com.example.desafioandroid.model.Repository;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RepositoryActivity extends AppCompatActivity implements RepositoryAdapter.RepositoryAdapterOnClickListener {

    //Views
    private RecyclerView mRecyclerView;

    //ViewModel
    private RepositoryViewModel viewModel;

    //Adapter
    private RepositoryAdapter mRepositoryAdapter;

    //Sercice
    private GetDataService service;

    //onScroll
    boolean isScrolling = false;
    int currentItems, totalItens, scrollOutItems;
    private int mCurrentPage;

    //Anothers Variables
    private static final String CURRENT_PAGE_KEY = "CURRENT_PAGE";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_repository);

        this.mRecyclerView = findViewById(R.id.recyclerViewRepository);

        if (savedInstanceState != null){
            mCurrentPage =savedInstanceState.getInt(CURRENT_PAGE_KEY);
            Log.d("CURRENT_PAGE", "SALVO");
        }else{
            mCurrentPage = 1;
            Log.d("CURRENT_PAGE", "NAO SALVO");
        }

        initUI();
        setupViewModel();
    }

    private void setupViewModel() {

        viewModel = ViewModelProviders.of(this, new RepositoryViewModelFactory(this.getApplication(), mCurrentPage)).get(RepositoryViewModel.class);
        viewModel.observerBaseRepository().observe(this, new Observer<Base>() {
            @Override
            public void onChanged(Base base) {
                mRepositoryAdapter.updateItems(base.getRepositories());
            }

        });


    }

    private void initUI() {
        mRepositoryAdapter = new RepositoryAdapter(null, this, this);

        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        mRecyclerView.setHasFixedSize(true);

        mRecyclerView.setLayoutManager(linearLayoutManager);
        mRecyclerView.setAdapter(mRepositoryAdapter);

        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if (newState == AbsListView.OnScrollListener.SCROLL_STATE_TOUCH_SCROLL) {
                    isScrolling = true;
                }
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                currentItems = linearLayoutManager.getChildCount();
                totalItens = linearLayoutManager.getItemCount();
                scrollOutItems = linearLayoutManager.findFirstVisibleItemPosition();

                if (isScrolling && (currentItems + scrollOutItems == totalItens)) {
                    isScrolling = false;
                    fetchData();
                }
            }
        });
    }


    private void fetchData() {
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                mCurrentPage++;
                updateListRepository();
            }
        }, 1000);
    }


    private void updateListRepository() {
        service = RetrofitClientInstance.getRetrofitInstance().create(GetDataService.class);
        Call<Base> call = service.getBaseResponse(mCurrentPage);

        call.enqueue(new Callback<Base>() {
            @Override
            public void onResponse(Call<Base> call, Response<Base> response) {
                Base base = response.body();
                mRepositoryAdapter.updateItems2(base.getRepositories());

            }

            @Override
            public void onFailure(Call<Base> call, Throwable t) {

            }
        });
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(CURRENT_PAGE_KEY, mCurrentPage);
    }

    @Override
    public void onMyClickListener(Repository repository) {
        Toast.makeText(this, repository.getName(), Toast.LENGTH_SHORT).show();
        Intent intent = new Intent(this, PullActivity.class);
        intent.putExtra("USER", repository.getOwner().getLogin());
        intent.putExtra("REPOSITORY", repository.getName());
        startActivity(intent);
    }

}

File: app/src/main/java/br/com/rmso/popularrepositories/ui/activities/RepositoryActivity.kt
package br.com.rmso.popularrepositories.ui.activities

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import br.com.rmso.popularrepositories.ListOnClickListener
import br.com.rmso.popularrepositories.R
import br.com.rmso.popularrepositories.model.Repository
import br.com.rmso.popularrepositories.model.RepositoryListCallback
import br.com.rmso.popularrepositories.retrofit.RetrofitAPI
import br.com.rmso.popularrepositories.ui.adapters.RepositoryAdapter
import br.com.rmso.popularrepositories.utils.Constants
import kotlinx.android.synthetic.main.activity_repository.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class RepositoryActivity : AppCompatActivity(), ListOnClickListener {
    private var retrofit = RetrofitAPI()
    private var page = 1
    private var repositoriesArrayList = ArrayList<Repository>()
    private var isLoading = false
    private var lastPosition = 0
    private val constants = Constants()

    var linearLayoutManager = LinearLayoutManager(this@RepositoryActivity)
    var adapterRepository = RepositoryAdapter(repositoriesArrayList, this@RepositoryActivity)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_repository)

        requestList()

        rv_repository.apply {
            setHasFixedSize(true)
            layoutManager = linearLayoutManager
            adapter = adapterRepository

            updateList(repositoriesArrayList,lastPosition)

            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                    val linearLayoutManager = recyclerView.layoutManager as LinearLayoutManager
                    val lastCompleteItem = linearLayoutManager.findLastCompletelyVisibleItemPosition()
                    if (!isLoading) {
                        if (lastCompleteItem == repositoriesArrayList.size - 1){
                            page += 1
                            isLoading = true
                            requestList()
                        }
                    }
                }
            })
        }
    }

    private fun updateList(list: ArrayList<Repository>, lastPosition: Int){
        if (page > 1) {
            rv_repository.adapter?.notifyItemRangeInserted(lastPosition, list.size)
        }else {
            rv_repository.adapter?.notifyDataSetChanged()
        }
        isLoading = false
    }

    fun requestList() {
        setProgressBar(true)
        val callRespository = retrofit.repositoryService.listRepositories(page)
        callRespository.enqueue(object : Callback<RepositoryListCallback> {
            override fun onResponse(call: Call<RepositoryListCallback>, response: Response<RepositoryListCallback>) {
                val listRepositories = response.body()!!
                setProgressBar(false)
                lastPosition = repositoriesArrayList.size + 1
                repositoriesArrayList.addAll(listRepositories.items)
                updateList(repositoriesArrayList,lastPosition)

            }

            override fun onFailure(call: Call<RepositoryListCallback>, t: Throwable) {
                Log.e(constants.msgError, t.message)
                setProgressBar(false)
            }
        })
    }

    override fun onClick(position: Int) {
        val intent = Intent(this@RepositoryActivity, PullRequestActivity::class.java)
        val repository = repositoriesArrayList[position]
        intent.putExtra(constants.owner, repository.owner.login)
        intent.putExtra(constants.repository, repository.name)
        startActivity(intent)
    }

    fun setProgressBar(status: Boolean){
        if(status) {
            pb_loading.visibility = View.VISIBLE
        }else {
            pb_loading.visibility = View.GONE
        }
    }
}
